src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -575,49 +589,59 @@ private bool DelegateNeedsReturn(MethodSymbol? invokeMethod)
-                Binder,
-                compilation,
-                Binder.ContainingMemberOrLambda,
-                _unboundLambda,
-                parameterTypes,
-                parameterRefKinds,
-                refKind: default,
-                returnType: default);
-            var lambdaBodyBinder = new ExecutableCodeBinder(_unboundLambda.Syntax, lambdaSymbol, ParameterBinder(lambdaSymbol, Binder));
-            var block = BindLambdaBody(lambdaSymbol, lambdaBodyBinder, BindingDiagnosticBag.Discarded);
-            var returnTypes = ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)>.GetInstance();
-            BoundLambda.BlockReturns.GetReturnTypes(returnTypes, block);
-            var inferredReturnType = BoundLambda.InferReturnType(
-                returnTypes,
-                _unboundLambda,
-                lambdaBodyBinder,
-                delegateType: null,
-                isAsync: IsAsync,
-                Binder.Conversions);
-
-            if (HasExplicitlyTypedParameterList && (inferredReturnType.TypeWithAnnotations.HasType || returnTypes.Count == 0))
+            if (!HasExplicitReturnType(out var returnRefKind, out var returnType))
             {
-                return Binder.GetMethodGroupOrLambdaDelegateType(
-                    inferredReturnType.RefKind,
-                    inferredReturnType.TypeWithAnnotations,
-                    parameterRefKinds,
+                var lambdaSymbol = new LambdaSymbol(
+                    Binder,
+                    Binder.Compilation,
+                    Binder.ContainingMemberOrLambda,
+                    _unboundLambda,
                     parameterTypes,
-                    ref useSiteInfo);
+                    parameterRefKinds,
+                    refKind: default,
+                    returnType: default);
+                var lambdaBodyBinder = new ExecutableCodeBinder(_unboundLambda.Syntax, lambdaSymbol, ParameterBinder(lambdaSymbol, Binder));
+                var block = BindLambdaBody(lambdaSymbol, lambdaBodyBinder, BindingDiagnosticBag.Discarded);
+                var returnTypes = ArrayBuilder<(BoundReturnStatement, TypeWithAnnotations)>.GetInstance();
+                BoundLambda.BlockReturns.GetReturnTypes(returnTypes, block);
+                var inferredReturnType = BoundLambda.InferReturnType(
+                    returnTypes,
+                    _unboundLambda,
+                    lambdaBodyBinder,
+                    delegateType: null,
+                    isAsync: IsAsync,
+                    Binder.Conversions);
+
+                returnType = inferredReturnType.TypeWithAnnotations;
+                returnRefKind = inferredReturnType.RefKind;
+
+                if (!returnType.HasType && returnTypes.Count > 0)
+                {
+                    return null;
+                }
             }
 
-            return null;
+            return Binder.GetMethodGroupOrLambdaDelegateType(
```

Comment:
>GetMethodGroupOrLambdaDelegateType [](http://example.com/codeflow?start=26&length=34)

It looks like the other caller of `GetMethodGroupOrLambdaDelegateType` has some similar logic for handling `void`. Should we push this logic down into `GetMethodGroupOrLambdaDelegateType`?

Reply:
> Should we push this logic down into `GetMethodGroupOrLambdaDelegateType`?

I considered that but decided against it because that would require `GetMethodGroupOrLambdaDelegateType` to allow two ways to represent `System.Void` - either `null` or an actual `TypeSymbol` - because `InferReturnType()` will return `null` for that case.